### DISCUSSION TITLE
Cookies for API & Middleware working in Dev but not Live

### DISCUSSION THREAD

**Discussion by dev-fredericfox** [nuxt/nuxt#17906] [Questions] [answered]
Hey everyone,

I have an issue understanding why my cookies work locally but not on a live server. Can anyone help me understand what I am doing wrong?

**Live**: https://nuxt3-middleware-issue.web.app/
**GitHub**: https://github.com/dev-fredericfox/nuxt3-middleware-issue

As you can see on the live preview on the second and third line the cookie is undefined. However when I test this locally it works fine.
**Local**:
<img width="679" alt="Screenshot 2022-02-23 at 11 54 11" src="https://user-images.githubusercontent.com/85335421/155305999-c7bf6310-9de3-4515-ba31-6344a9d01417.png">

**Live**:
![Screenshot 2022-02-23 at 12 03 24](https://user-images.githubusercontent.com/85335421/155307232-99aa854f-b64a-4056-bf66-b62f26943914.png)

Thank you!

**Comment by danielroe** [comment, marked as answer]
Check out https://v3.nuxtjs.org/docs/usage/data-fetching#isomorphic-fetch.

**Reply by dev-fredericfox** [reply]
@danielroe Thank you for the answer, I changed my fetch to:

```javascript
// pages/index.vue
const {data: publicEndpoint} = await useLazyFetch('/api/publicEndpoint', {
    headers: useRequestHeaders(['cookie'])
})
const {data: privateEndpoint} = await useLazyFetch('/api/privateEndpoint', {
    headers: useRequestHeaders(['cookie'])
})
```

Also updated the firebase code, and again on Localhost it works (dev and build), but it's still not working on firebase. 


I also added some logging, in the firebase functions, but it still seems to be undefined.

```javascript
// server/middleware/admin.ts
import * as functions from 'firebase-functions';
import type { IncomingMessage, ServerResponse } from 'http'
import { useCookie } from 'h3';

export default async (req: IncomingMessage, res: ServerResponse,next) => {
    if (req.url.includes('/api/privateEndpoint')) {
    const token = useCookie(req, 'test');
    const message = `I am the middleware. My received cookie is: ${token}`
    
    console.log(req.headers.cookie)
    console.log(token)
    functions.logger.log('header cookie',req.headers.cookie);
    functions.logger.log('Token:',token);

    res.end(message)
    }
    next()
}
```
```javascript
// server/api/publicEndpoint.ts
import { useCookies, useQuery } from 'h3'
import * as functions from 'firebase-functions';

export default async (req, res) => {
  const cookies = useCookies(req)
  functions.logger.log("cookies:",cookies);
  const message = "I am the public route"
  return { message, cookies }
}
```


![Screenshot 2022-02-23 at 15 10 32](https://user-images.githubusercontent.com/85335421/155335874-2bac078e-c7de-409b-be70-385ff2d93289.png)

Any more ideas what I am doing wrong?

**Reply by danielroe** [reply]
I would check that your cookies are defined correctly - look into secure/httpOnly/SameSite cookie options. (This is likely the issue if the firebase local preview works.) If not, then it may be that firebase has a provider-specific way to expose cookies which we need to implement in Nitro.

**Reply by dev-fredericfox** [reply]
@danielroe 
I changed the live site and updated the github repo, trying many different combinations of secure/httpOnly/SameSite cookie options, still no luck. If you have any ideas of other things to try, I would be grateful.
https://nuxt3-middleware-issue.web.app/
Everything works as expected in firebase local preview, but not on firebase live.

**Reply by DevJoghurt** [reply]
If you are using Firebase Hosting + Cloud Functions, __session is the only cookie you can use. Cookies are generally stripped from incoming requests. Firebase uses this behaviour for CDN caching.
You can find more info here: 
https://firebase.google.com/docs/hosting/manage-cache

**Reply by dev-fredericfox** [reply]
Ah thank you so much! Very interesting.

Indeed `__session` seems to be the only cookie that works, tested with Nuxt 3.
